engine/src/main/java/org/terasology/context/internal/ContextImpl.java

```diff
@@ -54,12 +53,51 @@ public ContextImpl() {
         if (parent != null) {
             return parent.get(type);
         }
-        return result;
+        return null;
     }
 
     @Override
-    public <T, U extends T> void put(Class<T> type, U object)  {
+    public <T, U extends T> void put(Class<T> type, U object) {
         map.put(type, object);
     }
 
+    // DynamicInstanceProvider stuff, kinda messy but it works
+
+    private final Map<Class<?>, DynamicInstanceProviderHolder<?>> providers = Maps.newConcurrentMap();
+
+    @Override
+    public <T, U extends T> void putInstanceProvider(Class<T> type, DynamicInstanceProvider<U> provider) {
+        providers.put(type, new DynamicInstanceProviderHolder<>(provider.getClass(), provider));
+    }
+
+    @Override
+    public <T> DynamicInstanceProvider<T> getInstanceProvider(Class<? extends T> type) {
+        if (System.getSecurityManager() != null) {
+            System.getSecurityManager().checkPermission(new RuntimePermission("permGetInstanceProvider"));
+        }
```

Comment:
I would like to keep the Context interface and the injection concept as simple as possible.

Some idea I could think of to simplify it a little bit would be to have `Context` have a `inject` method so that we don't need to have a method that we must protect against all but one caller. Although I am not that happy with that idea.